File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        ## create two hashsets/dictionary for finding duplicates , both for rows and columns
        
        cols = collections.defaultdict(set)
        rows = collections.defaultdict(set)
        
        ## also create a hashset for each 3*3 grid
        
        squares = collections.defaultdict(set) # key = (r/3 , c/3)
        
        for r in range(9):
            for c in range(9):
                ## if the value is "." move to the next iteration
                if(board[r][c] == "."):
                    continue
                  
                if(board[r][c] in rows[r] or board[r][c] in cols[c] or board[r][c] in squares[(r//3 , c//3)]):
                    return False
                ## and if no duplicates then add values into the board
                
                cols[c].add(board[r][c])
                rows[r].add(board[r][c])
                squares[(r//3 , c//3)].add(board[r][c])
              
                        
                        
        return True             
        

File: 0036-valid-sudoku/0036-valid-sudoku.cpp
//space : o(9) or o(1)
//  time: o(81) or o(1)
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int, unordered_set<char>> cols;
        unordered_map<int, unordered_set<char>> rows;
        unordered_map<int, unordered_set<char>> squares;

        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                if (board[r][c] == '.') {
                    continue;
                }

                if (rows[r].find(board[r][c]) != rows[r].end() ||
                    cols[c].find(board[r][c]) != cols[c].end() ||
                    squares[3 * (r / 3) + c / 3].find(board[r][c]) != squares[3 * (r / 3) + c / 3].end()) {
                    return false;
                }

                cols[c].insert(board[r][c]);
                rows[r].insert(board[r][c]);
                squares[3 * (r / 3) + c / 3].insert(board[r][c]);
            }
        }
        
        return true;
    }
};

/*
int main() {
    Solution solution;
    vector<vector<char>> input_board = {
        {'5','3','.','.','7','.','.','.','.'},
        {'6','.','.','1','9','5','.','.','.'},
        {'.','9','8','.','.','.','.','6','.'},
        {'8','.','.','.','6','.','.','.','3'},
        {'4','.','.','8','.','3','.','.','1'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','6','.','.','.','.','2','8','.'},
        {'.','.','.','4','1','9','.','.','5'},
        {'.','.','.','.','8','.','.','7','9'}
    };

    bool result = solution.isValidSudoku(input_board);

    cout << (result ? "Valid Sudoku" : "Invalid Sudoku") << endl;

    return 0;
}
*/
